### PR TITLE
utilities for HGCAL end of life

### PR DESCRIPTION
**Pull request by franzoni**
#### PR description:

Only technical changes:
 - avoid looping over in-time and out-of-time HGCAL digis if a data farme wont' be kept
- add utilities and customise functions we've used to validate the HGCAL noise model and occupancies toggling on-and-off noise, out-of-time
- implement interface to actually use the existing capability of scaling the fluence map to model HGCAL ageing at an arbitrary integrated luminosity

#### PR validation:

We've reported on occupancy and spectra on slides 17-22 of this [talk at the HGCAL mini-workshop last week](https://indico.cern.ch/event/933714/contributions/3924245/attachments/2072440/)

Not a backport.

@pfs @rovere

### CODE REVIEW

**Comment by franzoni** [comment]
please test

**Comment by franzoni** [comment]
[being addressed.](https://github.com/cms-sw/cmssw/pull/30697#issuecomment-658739258)

**Comment by franzoni** [comment]
I've added one commit to the branch [hgcal_eol_pulse_update_112X_six](https://github.com/franzoni/cmssw/tree/hgcal_eol_pulse_update_112X_six) however somehow the new commit is not being picked up by the PR code changes.
I'll wait a bit more and re-trigger the tests later.

**Comment by franzoni** [comment]
The commit eventually got taken into consideration.
It addresses the 3 issues previously identified by the relval tests.
I'll re-trigger the tests.

**Comment by franzoni** [comment]
please test

**Comment by kpedro88** [comment]
+upgrade

File: SimCalorimetry/HGCalSimAlgos/test/HGCSiNoiseMapAnalyzer.cc [hunk 1]
```diff
@@ -82,11 +84,15 @@ HGCSiNoiseMapAnalyzer::HGCSiNoiseMapAnalyzer(const edm::ParameterSet &iConfig) {
 
   noiseMaps_[DetId::HGCalEE] = std::unique_ptr<HGCalSiNoiseMap>(new HGCalSiNoiseMap);
   noiseMaps_[DetId::HGCalEE]->setDoseMap(doseMapURL, doseMapAlgo);
+  noiseMaps_[DetId::HGCalEE]->setDoseMap(doseMapURL, doseMapAlgo);
```
Comment (perrotta) [inline SimCalorimetry/HGCalSimAlgos/test/HGCSiNoiseMapAnalyzer.cc:87]: Duplicate line

File: RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h [hunk 2]
```diff
@@ -104,6 +104,7 @@ class HGCalImagingAlgo : public HGCalClusteringAlgoBase {
     iDesc.add<edm::ParameterSetDescription>("noises", descNestedNoises);
     edm::ParameterSetDescription descNestedNoiseMIP;
     descNestedNoiseMIP.add<bool>("scaleByDose", false);
+    descNestedNoiseMIP.add<double>("scaleByDoseFactor", 1.);
```
Comment (perrotta) [inline RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h:107]: Could you please remind me why all these parameters are defined inside a Reco producer, but they are only read and used in Simulation?

Comment (franzoni) [reply RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h:107]: Ciao @perrotta valid and good question.
The cluster algorithms import, by deliberate choice, parameters from the digitisation psets to access info like noise etc, and to ensure they use values consistent with those employed upstream. Together with parameters actually employed, the descriptor of the clustering producers end up having to include also  values which are not employed actually.
I don't particularly like this set up however I would not plan to re-engineer this aspect of the clustering and if these lines are not added redo crashes lamenting the producer is provided an invalid parameter.

I hope I am clear and accurate enough.

File: SimCalorimetry/HGCalSimAlgos/test/HGCSiNoiseMapAnalyzer.cc (hunk 1 above)
Comment (franzoni) [reply SimCalorimetry/HGCalSimAlgos/test/HGCSiNoiseMapAnalyzer.cc:87]: Thanks will address once back to a better keyboard

Comment (franzoni) [reply SimCalorimetry/HGCalSimAlgos/test/HGCSiNoiseMapAnalyzer.cc]: done.

**Comment by franzoni** [comment]
please test

**Comment by franzoni** [comment]
please test

**Comment by perrotta** [comment]
+1
- Changes in reco are consequent to the technical updates in the simulation part
- Having the simulation parameters defined inside the reco producers is not satisfactory in the long term, but it is not the scope of this PR to fix it
- Jenkins tests pass and show no differences, as expected

**Comment by civanch** [comment]
+1

**Comment by kpedro88** [comment]
+upgrade

**Comment by silviodonato** [comment]
+1

File: RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h (hunk 2 above)
Comment (rovere) [reply RecoLocalCalo/HGCalRecProducers/interface/HGCalImagingAlgo.h:107]: I think the problem is not related to the underlying mechanism of validating the input parameters but to the way in which parameters used for reconstruction and simulation have been bundled together in the same PSet. Once you've made this choice, anything else is just a mere consequence of that.

**Pull request merged**